README.md
# async_button_builder

AsyncButtonBuilder offers a simple way to extend any type of button with an asynchronous aspect. It allows adding loading, disabled, errored and completed states (with fluid animation between each) on top of buttons that perform asynchronous tasks.

## Getting Started

Include the package:

```yaml
  async_button_builder: <latest_version>
```

Wrap the builder around a button, passing the onPressed and child element to builder instead of the button directly. These two are the only required fields.

```dart
AsyncButtonBuilder(
  child: Text('Click Me'),
  onPressed: () async {
    await Future.delayed(Duration(seconds: 1));
  },
  builder: (context, child, callback, _) {
    return TextButton(
      child: child,
      onPressed: callback,
    );
  },
),
```

<p>  
 <img src="https://github.com/Nolence/async_button_builder/blob/main/screenshots/ezgif-7-61c436edaec2.gif?raw=true"/>
</p>

The fourth value in the builder allows you listen to the loading state. This can be used to conditionally style the button. This package depends `freezed` in order to create a sealed union to better handle the possible states.

> NOTE (Breaking change): As of v3.0.0, error now takes the error and stack trace as arguments.

```dart
AsyncButtonBuilder(
  child: Text('Click Me'),
  loadingWidget: Text('Loading...'),
  onPressed: () async {
    await Future.delayed(Duration(seconds: 1));

    // See the examples file for a way to handle timeouts
    throw 'yikes';
  },
  builder: (context, child, callback, buttonState) {
    final buttonColor = buttonState.when(
      idle: () => Colors.yellow[200],
      loading: () => Colors.grey,
      success: () => Colors.orangeAccent,
      error: (err, stack) => Colors.orange,
    );

    return OutlinedButton(
      child: child,
      onPressed: callback,
      style: OutlinedButton.styleFrom(
        primary: Colors.black,
        backgroundColor: buttonColor,
      ),
    );
  },
),
```

<p>  
 <img src="https://github.com/Nolence/async_button_builder/blob/main/screenshots/ezgif-7-a971c6afaabf.gif?raw=true"/>
</p>

You can also drive the state of the button yourself using the  `buttonState` field:

```dart
AsyncButtonBuilder(
  buttonState: ButtonState.completing(),
  // ...
),
```

## Notifications

As of v3.0.0, you can now wrap a higher level parent to handle notifications that come from buttons. Why not use something like `runZonedGuarded`? Notification bubbling handles not only the error but the state of the button. If you'd like, for example, to trigger a circular spinner in the center of the app notifiying the user that something is happening, you can do so by listening to the `AsyncButtonNotification` and then using the `buttonState` to determine what to do.

It might also be a good idea to separate the errors that come from button presses and those that are not. An error wants to see why a button press silently failed but might not need to know why a background fetch failed.

```dart
MaterialApp(
  home: NotificationListener<AsyncButtonNotification>(
    onNotification: (notification) {
      notification.buttonState.when(
        idle: () => // nothing -> you could use a maybeWhen as well
        loading: () => // show circular loading widget?
        success: () => // show success snackbar?
        error: (_, __) => // show error snackbar?
      );

      // Tells the notification to stop bubbling
      return true;
    },
    // This async button can be nested arbitrarily deep*
    child: AsyncButtonBuilder(
      duration: duration,
      errorDuration: const Duration(milliseconds: 100),
      errorWidget: const Text('error'),
      onPressed: () async {
        throw ArgumentError();
      },
      builder: (context, child, callback, state) {
        return TextButton(onPressed: callback, child: child);
      },
      child: const Text('click me'),
    ),
  ),
)

// See NotificationListener for more information
```

To disable the notifications, you can pass `false` to `notifications`.

## Customization

`async_button_builder` even works for custom buttons. You can define your own widgets for loading, error, and completion as well as define the transitions between them. This example is a little verbose but shows some of what's possible.


```dart
AsyncButtonBuilder(
  child: Padding(
    padding: const EdgeInsets.symmetric(
      horizontal: 16.0,
      vertical: 8.0,
    ),
    child: Text(
      'Click Me',
      style: TextStyle(color: Colors.white),
    ),
  ),
  loadingWidget: Padding(
    padding: const EdgeInsets.all(8.0),
    child: SizedBox(
      height: 16.0,
      width: 16.0,
      child: CircularProgressIndicator(
        valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
      ),
    ),
  ),
  successWidget: Padding(
    padding: const EdgeInsets.all(4.0),
    child: Icon(
      Icons.check,
      color: Colors.purpleAccent,
    ),
  ),
  onPressed: () async {
    await Future.delayed(Duration(seconds: 2));
  },
  loadingSwitchInCurve: Curves.bounceInOut,
  loadingTransitionBuilder: (child, animation) {
    return SlideTransition(
      position: Tween<Offset>(
        begin: Offset(0, 1.0),
        end: Offset(0, 0),
      ).animate(animation),
      child: child,
    );
  },
  builder: (context, child, callback, state) {
    return Material(
      color: state.maybeWhen(
        success: () => Colors.purple[100],
        orElse: () => Colors.blue,
      ),
      // This prevents the loading indicator showing below the
      // button
      clipBehavior: Clip.hardEdge,
      shape: StadiumBorder(),
      child: InkWell(
        child: child,
        onTap: callback,
      ),
    );
  },
),
```

<p>  
 <img src="https://github.com/Nolence/async_button_builder/blob/main/screenshots/ezgif-7-4088c909ba83.gif?raw=true"/>
</p>

Issues and PR's welcome

CHANGELOG.md
## [3.0.0+1] - 2022-02-13
- Fix `Undefined name 'optionalTypeArgs'.` Also adds back the hard dep on freezed_annotation.

## [3.0.0] - 2022-02-13
- Added notification functionality. The error variant now passes through the error and stacktrace which would be needed for a notification to be useful. This is a breaking change as the error widget now takes an error and stacktrace as arguments.

## [2.3.0+2] - 2023-02-13
- Not pushed

## [2.3.0+1] - 2022-05-24
- Added back freezed_annotation as a hard dep.

## [2.3.0+1] - 2022-05-24
- Fixed onError never being called when an error occurs - Credit @jonjomckay

## [2.3.0] - 2022-05-24
- Removed freezed dependency and added new callbacks @mergehez
- Upgrade to freezed 2 (excluding user now) @quantosapplications @esenmx

## [2.2.0+1] - 2022-01-28
Ignore .vscode, update versions and re-export button state. Credit @esenmx

## [2.2.0] - 2022-01-07
Got rid of hook version as it was complicating a simple package and likely discouraging possible contributors.

## [2.1.4] - 2021-09-28
Fix manual button state changes. Credit to @iliser

## [2.1.3+9] - 2021-09-28
Upgraded deps

## [2.1.3+8] - 2021-09-28
Upgraded deps

## [2.1.3+7] - 2021-09-08
Fixed github actions

## [2.1.3+6] - 2021-08-31
Trying to fix pub stuff

## [2.1.3+5] - 2021-08-27
Trying to fix pub stuff

## [2.1.3+4] - 2021-08-27
Fix home page

## [2.1.3+3] - 2021-08-12
Have proper types exported

## [2.1.3+2] - 2021-08-12
Fix possibility of nested async button builders conflicting with each other due to matching keyed subtrees. In order to avoid this situation, the parent key of the async_button_builder is used in the creation of sub widgets.

## [2.1.3+1] - 2021-06-25
Was not using proper sucessDuration when setting timeouts (regression)

## [2.1.3] - 2021-06-18
Tests and builds against stable. flutter_lints now used in place of pedantic

## [2.1.2+1] - 2021-06-12
Fixes analysis errors so I stop getting annoying emails

## [2.1.2] - 2021-05-24
Errors from onpressed will throw with orignal call stack

## [2.1.1+2] - 2021-04-14
Makes onpressed nullable similar to other button's behaviour

## [2.1.1+1] - 2021-04-14
Add documentation on how to handle timeouts

## [2.1.1] - 2021-04-14
ValueKeys are now no longer required in order to differentiate children

## [2.1.0] - 2021-02-12
Remove null-safety prefix as it's now in stable

## [2.0.9-nullsafety.0] - 2021-02-12
Trying to bump version again to go over stable

## [2.0.8-nullsafety.0] - 2021-02-12
Trying to bump version again to go over stable

## [2.0.7-nullsafety.8] - 2021-02-12
Republish so current nullsafe becomes main. My mistake

## [2.0.2-nullsafety.7] - 2021-02-12
Added a typedef in place of inline type on function. Added doc to `builder` field

## [2.0.2-nullsafety.6] - 2021-02-09
Renamed Union classes to match that of factory constructor name (Should not be breaking).

## [2.0.2-nullsafety.5] - 2021-02-01
Removed example integration test

## [2.0.2-nullsafety.4] - 2021-01-26
Adds tests and test coverage. Adds on dispose to cancel timer.

## [2.0.2-nullsafety.3] - 2021-01-25
Removes dangling controller and cleans up main code up a bit

## [2.0.2-nullsafety.2] - 2021-01-25
Fixes wrong image path

## [2.0.1-nullsafety.2] - 2021-01-25
Renames fields to better match standards of other buttons. 

## [2.0.0-dev.1] - 2021-01-25

Breaking. Replaces the fourth argument with a sealed union that better allows for directly managing states, removes unnecessary arguments such as padding, adds transition builders for custom transitions. Currently still includes AnimatedSize.

## [1.0.0-nullsafety.0] - 2021-01-12

Breaking. Adds a fourth argument of loading to the builder. This removes the value notifier that was difficult or confusing to manage in actual usage in my use cases. The argument isLoading is now a `bool`. This also adds a `disabled` field in order to set that on construction.

## [0.1.0-nullsafety.0] - 2021-01-12

Replaces the bool type of isLoading for a ValueNotifier<bool>.

## [0.0.1-nullsafety.2] - 2021-01-12

Adds analysis options and removes unnecessary typedef

## [0.0.1-nullsafety.1] - 2021-01-12

Adds docs to parameters and removes use of unnecessary undescore internal variables.

## [0.0.1-nullsafety.0] - 2021-01-11

Holds basic builder AsyncButtonBuilder. No tests yet. NNBD is supported


test/async_button_builder_test.dart
import 'package:async_button_builder/async_button_builder.dart';
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

void main() {
  testWidgets('displays child text', (WidgetTester tester) async {
    final textButton = MaterialApp(
      home: AsyncButtonBuilder(
        onPressed: () async {
          await Future<dynamic>.delayed(const Duration(seconds: 1));
        },
        builder: (context, child, callback, state) {
          return TextButton(onPressed: callback, child: child);
        },
        child: const Text('click me'),
      ),
    );

    await tester.pumpWidget(textButton);

    expect(find.text('click me'), findsOneWidget);
  });

  testWidgets('shows loading widget', (WidgetTester tester) async {
    const duration = Duration(milliseconds: 250);
    final textButton = MaterialApp(
      home: AsyncButtonBuilder(
        duration: duration,
        loadingWidget: const Text('loading'),
        onPressed: () async {
          await Future<dynamic>.delayed(const Duration(seconds: 1));
        },
        builder: (context, child, callback, state) {
          return TextButton(onPressed: callback, child: child);
        },
        child: const Text('click me'),
      ),
    );

    await tester.pumpWidget(textButton);

    await tester.tap(find.byType(TextButton));

    // 1/10 of a second later, loading should be showing
    await tester.pump(const Duration(milliseconds: 100));

    expect(find.text('loading'), findsOneWidget);

    // Let the widget continue to settle otherwise I won't dispose timers
    // correctly. TODO: Explain why .900 is the magic number
    await tester.pumpAndSettle(const Duration(milliseconds: 900));
  });

  testWidgets('shows error widget', (WidgetTester tester) async {
    const duration = Duration(milliseconds: 250);
    final textButton = MaterialApp(
      home: AsyncButtonBuilder(
        duration: duration,
        errorWidget: const Text('error'),
        onPressed: () async {
          throw ArgumentError();
        },
        builder: (context, child, callback, state) {
          return TextButton(onPressed: callback, child: child);
        },
        child: const Text('click me'),
      ),
    );

    await tester.pumpWidget(textButton);
    final button =
        find.byType(TextButton).evaluate().first.widget as TextButton;

    expect(
      () => button.onPressed!.call(),
      throwsA(isA<ArgumentError>()),
    );

    await tester.pump(const Duration(milliseconds: 200));

    expect(find.text('error'), findsOneWidget);
  });

  testWidgets('error notification bubbles up', (WidgetTester tester) async {
    const duration = Duration(milliseconds: 250);
    var idleCount = 0;
    var loadingCount = 0;
    var successCount = 0;
    var errorCount = 0;

    final textButton = MaterialApp(
      home: NotificationListener<AsyncButtonNotification>(
        onNotification: (notification) {
          notification.buttonState.when(
            idle: () => idleCount += 1,
            loading: () => loadingCount += 1,
            success: () => successCount += 1,
            error: (_, __) => errorCount += 1,
          );

          return true;
        },
        child: AsyncButtonBuilder(
          duration: duration,
          errorDuration: const Duration(milliseconds: 100),
          errorWidget: const Text('error'),
          onPressed: () async {
            throw ArgumentError();
          },
          builder: (context, child, callback, state) {
            return TextButton(onPressed: callback, child: child);
          },
          child: const Text('click me'),
        ),
      ),
    );

    await tester.pumpWidget(textButton);
    final button =
        find.byType(TextButton).evaluate().first.widget as TextButton;

    expect(
      () => button.onPressed!.call(),
      throwsA(isA<ArgumentError>()),
    );

    await tester.pump(const Duration(milliseconds: 200));

    expect(errorCount, 1);
    expect(idleCount, 1);
    expect(loadingCount, 1);
    expect(successCount, 0);
  });

  testWidgets('Returns to child widget', (WidgetTester tester) async {
    const duration = Duration(milliseconds: 250);
    final textButton = MaterialApp(
      home: AsyncButtonBuilder(
        duration: duration,
        loadingWidget: const Text('loading'),
        onPressed: () async {
          await Future<dynamic>.delayed(const Duration(seconds: 1));
        },
        builder: (context, child, callback, state) {
          return TextButton(onPressed: callback, child: child);
        },
        child: const Text('click me'),
      ),
    );

    await tester.pumpWidget(textButton);

    await tester.tap(find.byType(TextButton));

    await tester.pump(const Duration(milliseconds: 1000));

    expect(find.text('loading'), findsNothing);

    expect(find.text('click me'), findsOneWidget);
  });

  // TODO: Make it work on dropdown buttons
  // AsyncButtonBuilder(
  //   child: Icon(Icons.arrow_upward),
  //   onPressed: (newValue) async {
  //     final oldValue = dropdownValue;

  //     setState(() {
  //       dropdownValue = newValue;
  //     });

  //     await Future.delayed(Duration(seconds: 1));

  //     try {
  //       if (Random().nextBool()) {
  //         throw 'yikes';
  //       }
  //     } catch (error) {
  //       setState(() {
  //         dropdownValue = oldValue;
  //       });
  //     }
  //   },
  //   builder: (context, child, callback, _) {
  //     return DropdownButton<String>(
  //       // icon: SizedBox(
  //       //   height: 16.0,
  //       //   width: 16.0,
  //       //   child: CircularProgressIndicator(),
  //       // ),
  //       onChanged: callback,
  //       items: ['one', 'two', 'three']
  //           .map((value) => DropdownMenuItem(
  //                 child: Text(value),
  //                 value: value,
  //               ))
  //           .toList(),
  //       value: dropdownValue,
  //     );
  //   },
  // ),
}


lib/src/button_state/button_state.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'button_state.freezed.dart';

/// This union class represents the state of the button in either a [Idling],
/// [Loading], [Success] or [Error] state. This can be considered a enum with extra
/// utilities for ease of use.
///
/// {@tool snippet}
///
/// ```dart
/// final buttonColor = buttonState.when(
///   idle: () => Colors.pink,
///   loading: () => Colors.blue,
///   success: () => Colors.green,
///   error: () => Colors.red,
/// );
/// ```
/// {@end-tool}
///
/// You can also disregard other states and handle only those you'd like using
/// the `.maybeWhen` syntax.
///
/// /// {@tool snippet}
///
/// ```dart
/// final buttonColor = buttonState.maybeWhen(
///   idle: () => Colors.pink,
///   orElse: () => Colors.red,
/// );
/// ```
/// {@end-tool}
@freezed
class ButtonState with _$ButtonState {
  const factory ButtonState.idle() = Idle;
  const factory ButtonState.loading() = Loading;
  const factory ButtonState.success() = Success;
  const factory ButtonState.error(Object error, [StackTrace? stackTrace]) =
      Error;
}


example/lib/main.dart
import 'package:async_button_builder/async_button_builder.dart';
import 'package:flutter/material.dart';

void main() => runApp(const MyApp());

class MyApp extends StatelessWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData.dark(),
      home: const MyHomePage(),
    );
  }
}

class MyHomePage extends StatelessWidget {
  const MyHomePage({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(title: const Text('Async Buttons')),
      body: SizedBox.expand(
        child: Column(
          crossAxisAlignment: CrossAxisAlignment.center,
          mainAxisAlignment: MainAxisAlignment.spaceAround,
          children: [
            const Divider(),
            const Text('Text Button:'),
            AsyncButtonBuilder(
              child: const Text('Click Me'),
              onPressed: () async {
                await Future.delayed(const Duration(seconds: 1));
              },
              builder: (context, child, callback, _) {
                return TextButton(
                  onPressed: callback,
                  child: child,
                );
              },
            ),
            const Divider(),
            const Text('Elevated Button:'),
            AsyncButtonBuilder(
              child: const Text('Click Me'),
              onPressed: () async {
                await Future.delayed(const Duration(seconds: 1));
              },
              builder: (context, child, callback, _) {
                return ElevatedButton(
                  onPressed: callback,
                  child: child,
                );
              },
            ),
            const Divider(),
            const Text('Custom Outlined Button (Error):'),
            AsyncButtonBuilder(
              loadingWidget: const Text('Loading...'),
              onPressed: () async {
                await Future.delayed(const Duration(seconds: 1));

                throw 'yikes';

                // If you want to add a timeout, use something similar to
                //
                // try {
                //   await Future.delayed(Duration(seconds: 1))
                //       .timeout(Duration(milliseconds: 500));
                // } on TimeoutException catch (_) {
                //   // Show a popup or something
                //   rethrow;
                // } on Exception catch (_) {
                //   // Show a dialog or something
                //   rethrow;
                // }
                //
                // We rethrow so that async_button_builder can handle the error
                // state
              },
              builder: (context, child, callback, buttonState) {
                final buttonColor = buttonState.when(
                  idle: () => Colors.yellow[200],
                  loading: () => Colors.grey,
                  success: () => Colors.orangeAccent,
                  error: (_, __) => Colors.orange,
                );

                return OutlinedButton(
                  onPressed: callback,
                  style: OutlinedButton.styleFrom(
                    foregroundColor: Colors.black,
                    backgroundColor: buttonColor,
                  ),
                  child: child,
                );
              },
              child: const Text('Click Me'),
            ),
            const Divider(),
            const Text('Custom Material Button:'),
            const SizedBox(height: 6.0),
            AsyncButtonBuilder(
              loadingWidget: const Padding(
                padding: EdgeInsets.all(8.0),
                child: SizedBox(
                  height: 16.0,
                  width: 16.0,
                  child: CircularProgressIndicator(
                    valueColor: AlwaysStoppedAnimation<Color>(Colors.white),
                  ),
                ),
              ),
              successWidget: const Padding(
                padding: EdgeInsets.all(4.0),
                child: Icon(
                  Icons.check,
                  color: Colors.purpleAccent,
                ),
              ),
              onPressed: () async {
                await Future.delayed(const Duration(seconds: 2));
              },
              loadingSwitchInCurve: Curves.bounceInOut,
              loadingTransitionBuilder: (child, animation) {
                return SlideTransition(
                  position: Tween<Offset>(
                    begin: const Offset(0, 1.0),
                    end: const Offset(0, 0),
                  ).animate(animation),
                  child: child,
                );
              },
              builder: (context, child, callback, state) {
                return Material(
                  color: state.maybeWhen(
                    success: () => Colors.purple[100],
                    orElse: () => Colors.blue,
                  ),
                  // This prevents the loading indicator showing below the
                  // button
                  clipBehavior: Clip.hardEdge,
                  shape: const StadiumBorder(),
                  child: InkWell(
                    onTap: callback,
                    child: child,
                  ),
                );
              },
              child: const Padding(
                padding: EdgeInsets.symmetric(
                  horizontal: 16.0,
                  vertical: 8.0,
                ),
                child: Text(
                  'Click Me',
                  style: TextStyle(color: Colors.white),
                ),
              ),
            ),
            const Divider(),
          ],
        ),
      ),
    );
  }
}
